x-pack/metricbeat/module/googlecloud/timeseries_metadata_collector.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package googlecloud
+
+import (
+	"context"
+	"time"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/pkg/errors"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// NewStackdriverCollectorInputData returns a ready to use MetadataCollectorInputData to be sent to Metadata collectors
+func NewStackdriverCollectorInputData(ts *monitoringpb.TimeSeries, projectID, zone string) *MetadataCollectorInputData {
+	return &MetadataCollectorInputData{
+		TimeSeries: ts,
+		ProjectID:  projectID,
+		Zone:       zone,
+	}
+}
+
+// NewStackdriverMetadataServiceForTimeSeries apart from having a long name takes a time series object to return the
+// Stackdriver canonical Metadata extractor
+func NewStackdriverMetadataServiceForTimeSeries(ts *monitoringpb.TimeSeries) MetadataService {
+	return &StackdriverTimeSeriesMetadataCollector{
+		timeSeries: ts,
+	}
+}
+
+// StackdriverTimeSeriesMetadataCollector is the implementation of MetadataCollecter to collect metrics from Stackdriver
+// common TimeSeries objects
+type StackdriverTimeSeriesMetadataCollector struct {
+	timeSeries *monitoringpb.TimeSeries
+}
+
+// Metadata parses a Timeseries object to return its metadata divided into "unkonwn" (first object) and ECS (second
+// object https://www.elastic.co/guide/en/ecs/master/index.html)
+func (s *StackdriverTimeSeriesMetadataCollector) Metadata(ctx context.Context, in *monitoringpb.TimeSeries) (common.MapStr, common.MapStr, error) {
+	m := common.MapStr{}
+
+	var availabilityZone, accountID string
+
+	if in.Resource != nil && in.Resource.Labels != nil {
+		availabilityZone = in.Resource.Labels[JSON_PATH_ECS_AVAILABILITY_ZONE]
+		accountID = in.Resource.Labels[JSON_PATH_ECS_ACCOUNT_ID]
+	}
+
+	ecs := common.MapStr{
+		ECS_CLOUD: common.MapStr{
+			ECS_CLOUD_AVAILABILITY_ZONE: availabilityZone,
+			ECS_CLOUD_ACCOUNT: common.MapStr{
+				ECS_CLOUD_ACCOUNT_ID: accountID,
+			},
+			ECS_CLOUD_PROVIDER: "googlecloud",
+		},
+	}
+
```

Comment:
As the rest of the code in this function seems to need `s.timeSeries` to be not nil, should we return here only with the ECS map?
```suggestion
if s.timeSeries == nil {
    return m, ecs, nil
}
```

Reply:
Fixed

Reply:
Just for the record, I seem to remember that we already talked about this and this was the expected behaviour, so I am resolving this thread.